Add east-west and IPv4 load balancing

tgraf opened the pull request:
Reworks the load balancing code and makes most of the code available for reuse.

Adds the following capabilities:
- Balancing between containers on the same node
- Balancing between containers on different nodes
- Balancing from and to arbitrary IPs
- IPv4 support (limited)

IPv4 limitations:
- This does not include any work to carry the reverse NAT index in the packet from node to node. Reverse translation with IPv4 currently only works on the same node.

Example config, L3 load balancer based on L4 hash:

```
SVC_IP6="f00d::1:1"
LB_PORT=0
REVNAT=222
sudo cilium lb create-services-map
sudo cilium lb update-service $SVC_IP6 $LB_PORT 0 2 $REVNAT :: $LB_PORT
sudo cilium lb update-service $SVC_IP6 $LB_PORT 1 2 $REVNAT $SERVER1_IP $LB_PORT
sudo cilium lb update-service $SVC_IP6 $LB_PORT 2 2 $REVNAT $SERVER2_IP $LB_PORT
sudo cilium lb create-rev-nat-map
sudo cilium lb update-rev-nat $REVNAT $SVC_IP6 $LB_PORT
```

mchalla commented on bpf/lbmap/lb.go:
```diff
@@ -139,11 +152,32 @@ func init() {
-				ArgsUsage: "<map file> <maptype> <ipv6 addr> <dport> <state> <count> [<lxc-id> <lxc-port> <node-id> ...]",
-				Action:    lbUpdateKey,
+				ArgsUsage: "<ipv6 addr> <port> <state> <count> <lxc-id> <node-id>",
+				Action:    lbUpdateService,
```
This help needs to be updated.

mchalla commented on bpf/bpf_lb.c:
```diff
@@ -31,153 +31,109 @@
+
+	if (!(svc = lb6_lookup_slave(skb, &key, slave)))
+		return DROP_NO_SERVICE;
+
```
So we always need to do this additional lookup, its definitely more cleaner and simpler to use and I cannot think of how to avoid it with the new scheme.

tgraf replied:
This is another good example where masked lookups in the form if key1 does not match, then apply mask and try again will make sense. Same as for conntrack.

mchalla commented on bpf/bpf_lxc.c:
```diff
@@ -161,9 +174,15 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
-		cilium_trace(skb, DBG_GENERIC, state, ret);
+	/* Service redirection must occur before conntrack to ensure that the
+	 * conntrack entry is created for the translated final endpoint. */
+	if (!skip_service && (svc = lb6_lookup_service(skb, &key)) != NULL) {
```
Isn't this happening after conntrack ? Or am I missing something ?

tgraf replied:
Good catch, it was before conntrack before so the comment became outdated. I think it makes sense to eventually do it before conntrack which would allow to see the actual backend service in the CT table at the source. The reverse translation becomes slightly more complex though.

mchalla commented on bpf/lbmap/lb.go:
```diff
@@ -435,129 +451,145 @@ func lbLookupKey(ctx *cli.Context) {
 		return
 	}
-	lbval.lxcCount = int32(count)
+	lbval.count = uint16(count)
```
I suppose the count is only needed for the master and not the slaves. So should we update the map count, only the master changes with the addition or deletion of a slave.

tgraf replied:
Right. We need to figure out the automation part of this anyway.

mchalla commented on bpf/lib/ipv6.h:
```diff
@@ -237,20 +237,6 @@ static inline void ipv6_set_lxc_id(union v6addr *addr, __u16 lxc_id)
-	p4 |= (state << 16);
-	addr->p4 = htonl(p4);
-}
-
```
We still need these apis for vxlan mode where the state needs to be now derived from the ipv6 target and copied to the tunnel metadata and zeroed in the address.

tgraf replied:
I removed them because the byte order conversion  should not happen in the fast path at all. So when we do the encap part, let's add it properly with the minimal instructions required.

mchalla replied:
Sounds good.

aanm commented on bpf/bpf_lb.c:
```diff
@@ -15,6 +15,22 @@
+ *              IP and optional destination port for every IPv4 and
+ *              IPv6 packet recevied. If a matching entry is found, the
+ *              destination address will be written to one of the
+ *              configures slaves. Optionall the destination port can be
```
s/Optionall/Optionally

aanm commented on bpf/lbmap/lb.go:
```diff
@@ -42,62 +43,122 @@ import (
+
+func servicesMap() *bpf.Map {
+	if ipv4 {
+		return bpf.NewMap("/sys/fs/bpf/tc/globals/cilium_lb4_services",
```
`"/sys/fs/bpf/tc/globals/cilium_lb4_services"` -> `common.BPFCiliumMaps + "/cilium_lb4_services"`

aanm commented on bpf/lbmap/lb.go:
```diff
@@ -42,62 +43,122 @@ import (
+			int(unsafe.Sizeof(LB4Service{})),
+			mapSize)
+	} else {
+		return bpf.NewMap("/sys/fs/bpf/tc/globals/cilium_lb6_services",
```
`"/sys/fs/bpf/tc/globals/cilium_lb6_services"` -> `common.BPFCiliumMaps + "/cilium_lb6_services"`

aanm commented on bpf/lbmap/lb.go:
```diff
@@ -42,62 +43,122 @@ import (
+
+func stateMap() *bpf.Map {
+	if ipv4 {
+		return bpf.NewMap("/sys/fs/bpf/tc/globals/cilium_lb4_state",
```
`"/sys/fs/bpf/tc/globals/cilium_lb4_state"` -> `common.BPFCiliumMaps + "/cilium_lb4_state"`

aanm commented on bpf/lbmap/lb.go:
```diff
@@ -42,62 +43,122 @@ import (
+			int(unsafe.Sizeof(LB4State{})),
+			mapSize)
+	} else {
+		return bpf.NewMap("/sys/fs/bpf/tc/globals/cilium_lb6_state",
```
`"/sys/fs/bpf/tc/globals/cilium_lb6_state"` -> `common.BPFCiliumMaps + "/cilium_lb6_state"`

aanm commented on bpf/lbmap/lbmap.go:
```diff
@@ -0,0 +1,620 @@
+	"github.com/cilium/cilium/common/addressing"
+	"github.com/cilium/cilium/common/bpf"
+	"github.com/cilium/cilium/common/types"
+	"github.com/op/go-logging"
```
`"github.com/op/go-logging"` It's one line above from what it should be.

aanm replied:
Bump

tgraf committed:
map_ctrl: Make tool generic so it can work with any map

Acked-by: André Martins <andre@cilium.io>
Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Do not convert sec_label byteorder in debug message

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
vagrant: Enable IPv4 by default

Required for loadbalancer tests. Intermediate solution until we
can enable/disable IPv4 at runtime

Acked-by: André Martins <andre@cilium.io>
Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Add east-west and IPv4 load balancing

Reworks the load balancing code and makes most of the code available
for reuse.

Adds the following capabilities:
 * Balancing between containers on the same node
 * Balancing between containers on different nodes
 * Balancing from and to arbitrary IPs
 * IPv4 support (limited)
 *
IPv4 limitations:
This does not include any work to carry the reverse NAT index in the
packet from node to node. Reverse translation with IPv4 currently only
works on the same node.

Example config, L3 load balancer based on L4 hash:
SVC_IP6="f00d::1:1"
LB_PORT=0
REVNAT=222
sudo cilium lb create-services-map
sudo cilium lb update-service $SVC_IP6 $LB_PORT 0 2 $REVNAT :: $LB_PORT
sudo cilium lb update-service $SVC_IP6 $LB_PORT 1 2 $REVNAT $SERVER1_IP $LB_PORT
sudo cilium lb update-service $SVC_IP6 $LB_PORT 2 2 $REVNAT $SERVER2_IP $LB_PORT
sudo cilium lb create-rev-nat-map
sudo cilium lb update-rev-nat $REVNAT $SVC_IP6 $LB_PORT

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Add policy handler to bpf_overlay

When using overlay mode, loadbalancer punting to stack requires a
policy program to be available even in encap mode.

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on contrib/vagrant/start.sh:
```diff
@@ -70,7 +70,7 @@ function write_footer() {
 sleep 2s
 sed -i '/exec/d' /etc/init/cilium-net-daemon.conf
-echo 'exec cilium -D daemon run -n ${ipv6_addr} ${ipv4_options}-t vxlan -c "${NODE_IP_BASE}${FIRST_IP_SUFFIX}:8500"' >> /etc/init/cilium-net-daemon.conf
+echo 'exec cilium -D daemon run -n ${ipv6_addr} ${ipv4_options}-t vxlan --ipv4 -c "${NODE_IP_BASE}${FIRST_IP_SUFFIX}:8500"' >> /etc/init/cilium-net-daemon.conf
```
Why not setting `IPV4=1` at the top of the file? That option is being verified at line 66-68

tgraf replied:
Good point, changing this.
